src/main/java/in/projecteka/consentmanager/common/CentralRegistry.java

```diff
@@ -25,8 +26,31 @@
     }
 
     public Mono<String> authenticate() {
-        return clientRegistryClient
-                .getTokenFor(properties.getClientId(), properties.getXAuthToken())
-                .map(session -> format("%s %s", session.getTokenType(), session.getAccessToken()));
+        var accessToken = accessTokenCache.getIfPresent("accessToken").switchIfEmpty(Mono.just(""));
+        var refreshToken = refreshTokenCache.getIfPresent("refreshToken").switchIfEmpty(Mono.just(""));
+
+        return accessToken.flatMap(access -> {
+            if (access.isEmpty()) {
+                return refreshToken.flatMap(refresh -> {
+                    if (refresh.isEmpty()) {
+                        return clientRegistryClient.getTokenFor(properties.getClientId(), properties.getXAuthToken())
+                                .flatMap(session -> accessTokenCache.put("accessToken", session.getAccessToken()).then(
+                                        refreshTokenCache.put("refreshToken", session.getRefreshToken()).then(
+                                                Mono.just(String.format("%s %s", session.getTokenType(), session.getAccessToken()))
+                                        )
+                                ));
+                    } else {
+                        return clientRegistryClient.getTokenWithRefreshToken(properties.getClientId(), properties.getXAuthToken(), refresh)
+                                .flatMap(session -> accessTokenCache.put("accessToken", session.getAccessToken()).then(
+                                        refreshTokenCache.put("refreshToken", session.getRefreshToken()).then(
+                                                Mono.just(String.format("%s %s", session.getTokenType(), session.getAccessToken()))
+                                        )));
+                    }
+                });
+            }
+            return Mono.just(String.format("%s %s", "Bearer", access));
```

Comment:
```suggestion
        Mono<String> tokenUsingSecret =
				clientRegistryClient.getTokenFor(properties.getClientId(), properties.getXAuthToken())
						.flatMap(session -> accessTokenCache.put("accessToken", session.getAccessToken())
								.then(refreshTokenCache.put("refreshToken", session.getRefreshToken()))
								.thenReturn(session.getAccessToken()));
		Mono<String> tokenUsingRefreshToken = refreshTokenCache.getIfPresent("refreshToken")
				.switchIfEmpty(Mono.defer(() ->tokenUsingSecret))
				.flatMap(token -> clientRegistryClient.getTokenWithRefreshToken(properties.getClientId(),
						properties.getXAuthToken(),
						token)
						.flatMap(session -> accessTokenCache.put("accessToken", session.getAccessToken())
								.then(refreshTokenCache.put("refreshToken", session.getRefreshToken()))
								.thenReturn(session.getAccessToken())));
		return accessTokenCache.getIfPresent("accessToken")
				// but ideally it should be like (if empty call tokenUsingSecret, if expired call tokenUsingRefreshToken)
				.switchIfEmpty(Mono.defer(() -> tokenUsingRefreshToken))
				.map(token -> String.format("%s %s", "Bearer", token));
```

`It should be like (if empty call tokenUsingSecret, if expired call tokenUsingRefreshToken)`